warehouse/tracker.go

```diff
@@ -0,0 +1,183 @@
+package warehouse
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/utils/timeutil"
+
+	"github.com/rudderlabs/rudder-server/config"
+	"github.com/rudderlabs/rudder-server/rruntime"
+	"github.com/rudderlabs/rudder-server/services/stats"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+	"github.com/rudderlabs/rudder-server/warehouse/internal/model"
+	warehouseutils "github.com/rudderlabs/rudder-server/warehouse/utils"
+)
+
+func (wh *HandleT) CronTracker(ctx context.Context) error {
+	for {
+		wh.configSubscriberLock.RLock()
+
+		ch := make(chan warehouseutils.Warehouse, 1)
+		wg := sync.WaitGroup{}
+		wg.Add(len(wh.warehouses))
+
+		for _, warehouse := range wh.warehouses {
+			warehouse := warehouse
+
+			rruntime.GoForWarehouse(func() {
+				defer wg.Done()
+				ch <- warehouse
+			})
+		}
+		wh.configSubscriberLock.RUnlock()
```

Comment:
I try to get my head around this, why do we spawn up go routines, and we need a channel here?

Couldn't we just copy the warehouses in a new slice and pass them to `wh.Track`?

Reply:
Yeah. Initially, I was also thinking along similar lines. But seeing the implementation for the mainLoop I got biased. Making the changes.